ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,167 @@
+    this.neighborFinder = neighborFinder;
+    this.anteMap = new HashMap<>();
+    this.distanceSortedPeers = new TreeSet<>(Comparator.comparing(Map.Entry::getValue));
+    this.outstandingRequestList = new ArrayList<>();
+    this.contactedInCurrentExecution = new ArrayList<>();
+  }
+
+  /**
+   * The lookup initiator starts by picking α closest nodes to the target it knows of. The initiator
+   * then sends concurrent FindNode packets to those nodes. α is a system-wide concurrency
+   * parameter, such as 3.
+   */
+  private void initiatePeerRefreshCycle(final List<Peer> peers) {
+    for (Peer peer : peers) {
+      if (!contactedInCurrentExecution.contains(peer.getId())) {
+        BytesValue peerId = peer.getId();
+        outstandingRequestList.add(new OutstandingRequest(peerId));
+        contactedInCurrentExecution.add(peerId);
+        neighborFinder.issueFindNodeRequest(peer);
+      }
+      // The lookup terminates when the initiator has queried
+      // and gotten responses from the k closest nodes it has seen.
+    }
+  }
+
+  void digestNeighboursPacket(
+      final NeighborsPacketData neighboursPacket, final BytesValue peerIdentifier) {
+    if (outstandingRequestList.contains(new OutstandingRequest(peerIdentifier))) {
+      List<Peer> receivedPeerList = neighboursPacket.getNodes();
+      for (Peer receivedPeer : receivedPeerList) {
+        if (!peerBlacklist.contains(receivedPeer)) {
+          bondingAgent.performBonding(receivedPeer);
+          anteMap.put(receivedPeer, distance(target, receivedPeer.getId()));
+        }
+      }
+      outstandingRequestList.remove(new OutstandingRequest(peerIdentifier));
+      performIteration();
+    }
+  }
+
+  private List<Peer> determineFindNodeCandidates() {
+    distanceSortedPeers.addAll(anteMap.entrySet());
+    List<Map.Entry<Peer, Integer>> sortedList = new ArrayList<>(distanceSortedPeers);
+    List<Peer> queryCandidates = new ArrayList<>();
+    for (int i = 0; i < CONCURRENT_REQUEST_LIMIT; i++) {
+      if (i >= sortedList.size()) {
+        break;
+      }
+      queryCandidates.add(sortedList.get(i).getKey());
+    }
+    return queryCandidates;
+  }
+
+  private void performIteration() {
+    if (outstandingRequestList.isEmpty()) {
+      List<Peer> queryCandidates = determineFindNodeCandidates();
+      initiatePeerRefreshCycle(queryCandidates);
+    }
+  }
+
+  void kickstartBootstrapPeers(final List<Peer> bootstrapPeers) {
+    for (Peer bootstrapPeer : bootstrapPeers) {
+      BytesValue peerId = bootstrapPeer.getId();
+      outstandingRequestList.add(new OutstandingRequest(peerId));
+      contactedInCurrentExecution.add(peerId);
+      bondingAgent.performBonding(bootstrapPeer);
+      neighborFinder.issueFindNodeRequest(bootstrapPeer);
+    }
+  }
+
+  static class OutstandingRequest {
+    Instant creation;
+    BytesValue peerId;
+
+    OutstandingRequest(final BytesValue peerId) {
+      this.creation = Instant.now();
```

Comment:
We shouldn't use `Instant.now()` directly because it makes it too difficult to test the time handling.  Instead we should inject a `java.time.Clock` instance - then in production we can use `Clock.systemUTC()` and in tests we can use `Clock.fixed`, a mock or any other implementation we want so we can control time.

Reply:
I added some mechanism for testing the expiration of the outstanding requests, but maybe it's not rigorous enough. I'm happy to add this in, I just wanted to get some feedback on how it might work with the way that the expiration is being tested atm, since this comment was prior to that being implemented.

Reply:
We definitely need to abstract the clock - we don't want to have to sleep for 60 seconds in the test as it slows down the build too much.  Having the scheduling of the periodic job be external to this class and an abstracted clock would let us just advance the clock and trigger another check without any sleeping.